src/jquery.ime.js

```diff
@@ -496,264 +575,177 @@
-			var i, childNodesCount;
-
-			if ( node.nodeType === Node.TEXT_NODE ) {
-				if ( !foundStart && node === range.startContainer ) {
-					start = charIndex + range.startOffset;
-					foundStart = true;
-				}
-
-				if ( foundStart && node === range.endContainer ) {
-					end = charIndex + range.endOffset;
-					foundEnd = true;
-				}
-
-				charIndex += node.length;
-			} else {
-				childNodesCount = node.childNodes.length;
-
-				for ( i = 0; i < childNodesCount; ++i ) {
-					traverseTextNodes( node.childNodes[i], range );
-					if ( foundEnd ) {
-						break;
-					}
-				}
-			}
+		if ( sel.rangeCount === 0 ) {
+			return null;
 		}
-
-		if ( sel.rangeCount ) {
-			traverseTextNodes( element, sel.getRangeAt( 0 ) );
-		}
-
-		return [ start, end ];
-	}
-
-	function setCaretPosition( $element, position ) {
-		var currentPosition,
-			startCorrection = 0,
-			endCorrection = 0,
-			element = $element[0];
-
-		setDivCaretPosition( element, position );
-		currentPosition = getDivCaretPosition( element );
-		// see Bug https://bugs.webkit.org/show_bug.cgi?id=66630
-		while ( position.start !== currentPosition[0] ) {
-			position.start -= 1; // go back one more position.
-			if ( position.start < 0 ) {
-				// never go beyond 0
-				break;
-			}
-			setDivCaretPosition( element, position );
-			currentPosition = getDivCaretPosition( element );
-			startCorrection += 1;
+		range = sel.getRangeAt( 0 );
+		if ( !this.$element[ 0 ].contains( range.commonAncestorContainer ) ) {
+			return null;
 		}
+		return range;
+	};
 
-		while ( position.end !== currentPosition[1] ) {
-			position.end += 1; // go forward one more position.
-			setDivCaretPosition( element, position );
-			currentPosition = getDivCaretPosition( element );
-			endCorrection += 1;
-			if ( endCorrection > 10 ) {
-				// XXX avoid rare case of infinite loop here.
-				break;
-			}
-		}
+	TextEntryFactory.static.singleton.register( ContentEditableEntry );
 
-		return [startCorrection, endCorrection];
-	}
+	/* Exports */
 
 	/**
-	 * Set the caret position in the div.
-	 * @param {Element} element The content editable div element
-	 * @param position
+	 * jQuery plugin ime
+	 * @param {Object} option
 	 */
-	function setDivCaretPosition( element, position ) {
-		var nextCharIndex,
-			charIndex = 0,
-			range = rangy.createRange(),
-			foundStart = false,
-			foundEnd = false;
-
-		range.collapseToPoint( element, 0 );
-
-		function traverseTextNodes( node ) {
-			var i, len;
-
-			if ( node.nodeType === 3 ) {
-				nextCharIndex = charIndex + node.length;
+	$.fn.ime = function ( option ) {
+		return this.each( function () {
+			var data,
+				$this = $( this ),
+				options = typeof option === 'object' && option;
 
-				if ( !foundStart && position.start >= charIndex && position.start <= nextCharIndex ) {
-					range.setStart( node, position.start - charIndex );
-					foundStart = true;
-				}
+			// Some exclusions: IME shouldn't be applied to textareas with
+			// these properties.
```

Comment:
I am wondering whether these checks can now goto canwrap methods of respective classes. readonly and disabled are applicable only to formfields, and not applicable to a contenteditable field. While noime is applicable to both.